src/jit/importer.cpp

```diff
@@ -4039,8 +4039,37 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
-            case NI_MathF_Round:
-            case NI_Math_Round:
+#ifdef FEATURE_HW_INTRINSICS
+            case NI_System_Math_FusedMultiplyAdd:
+            case NI_System_MathF_FusedMultiplyAdd:
+            {
+#ifdef _TARGET_XARCH_
+                if (compSupports(InstructionSet_FMA))
+                {
+                    // We are constructing a chain of intrinsics similar to:
+                    //    return FMA.MultiplyAddScalar(
+                    //        Vector128.CreateScalar(x),
+                    //        Vector128.CreateScalar(y),
+                    //        Vector128.CreateScalar(z)
+                    //    ).ToScalar();
+
+                    GenTree* op3 = gtNewSimdHWIntrinsicNode(TYP_SIMD16, impPopStack().val,
+                                                            NI_Base_Vector128_CreateScalarUnsafe, callType, 16);
+                    GenTree* op2 = gtNewSimdHWIntrinsicNode(TYP_SIMD16, impPopStack().val,
+                                                            NI_Base_Vector128_CreateScalarUnsafe, callType, 16);
+                    GenTree* op1 = gtNewSimdHWIntrinsicNode(TYP_SIMD16, impPopStack().val,
+                                                            NI_Base_Vector128_CreateScalarUnsafe, callType, 16);
+                    GenTree* res =
+                        gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op2, op3, NI_FMA_MultiplyAddScalar, callType, 16);
+                    retNode = gtNewSimdHWIntrinsicNode(callType, res, NI_Base_Vector128_ToScalar, callType, 16);
```

Comment:
On `float` and `double`, `CreateScalar` and `ToScalar` is zero-codegen helper, so in importer we can just have `return FMA.MultiplyAddScalar(x, y, z)`.

Reply:
They ideally generate no codegen. However, there are still cases when they will (such as when the value is in memory and we the instruction expects a `m128`).

So we still need to insert the `ToScalar` and `CreateScalarUnsafe` nodes to ensure that the appropriate widening/narrowing happens (and to ensure that all the downstream asserts still pass).

Reply:
I am okay with the current implementation if it can generate the optimal code. But if not, removing `CreateScalar` and `ToScalar` could be a workaround. In this case, vectors are just intermediate tmp results, so we may not need to care about widening/narrowing with scalar calltype.

Reply:
Whether or not this generates optimal code, it is the correct place to begin. I'm concerned that mixing expected type sizes of (even intrinsic) methods could result in problems. Let's consider doing so only if we find that code generation is suboptimal, and then we can start by looking at what causes the pessimization.